### ISSUE TITLE
FAQ save to pickup folder caveat for IIS SMTP Server

### ISSUE THREAD

**Issue by TheAtomicOption** [jstedfast/MailKit#948]
**Proposal:**
I propose adding a note to the FAQ about a major caveat when using MimeKit/MailKit with IIS SMTP Server and a pickup folder.

The [MailKit FAQ ](https://github.com/jstedfast/MailKit/blob/master/FAQ.md#SpecifiedPickupDirectory) has code that shows how to stream out to an EML file in a pickup directory. However for people transitioning from `System.Net.Mail` internally, but not yet able to leave IIS SMTP Server, they need to account for Microsoft's non-standard division of labor with regard to dot-stuffing, [detailed here](https://github.com/axuno/MailMergeLib/wiki/IIS-SMTP-Server). 

The only way I found to successfully handle this is to use a MemoryStream rather than a FileStream with `Mimekit.WriteTo(stream)`, set `stream.Position = 0` afterwards, and then modify it directly, [as in this SO post](https://stackoverflow.com/questions/17864254/modify-filestream) by looking for `\r\n.` and replacing that with `\r\n..`, (assuming all line endings in your emails are windows style which for us is true) before writing to disk. 

This isn't _ideal_ given that some not-quoted-printable line in another part of the stream might start with a `.`, and I'm not clear whether that'd be problematic, but unless an output mode were built into MimeKit specifically for IIS SMTP Server (which is obsolete anyway--and would encourage MS' bad behavior re:RFCs), this is the option we have. 

**Context**:
For various reasons I switched to MimeKit, but am not yet able to switch to MailKit from IIS SMTP. I used the FAQ's snippet for saving files to a pickup folder which worked well overall (it creates thousands of 0 length files if the disk is full, but ideally you shouldn't let the disk fill). 

**Final Code**
```
SaveToPickupFolder(MimeMessage message, string path) {
do
{
	// Note: this will require that you know where the specified pickup directory is.
	// GetFreeSpaceOfPathInBytes() is to check that we have disk space to avoid creating 
	// 0 length files as we get the same exception for that as when there's a filename collision
	if (File.Exists(path) || GetFreeSpaceOfPathInBytes(path) < (1024ul * 1024ul * 1024ul))
		continue;

	try
	{
		StringBuilder sb = new StringBuilder();
		using (var stream = new MemoryStream())
		{
			message.WriteTo(stream);

                        // IIS SMTP doesn't handle dot-stuffing, so we have to do it manually here:
			stream.Position = 0;
			using (BufferedStream bs = new BufferedStream(stream))
			using (StreamReader sr = new StreamReader(bs))
			{
				string line;
				while ((line = sr.ReadLine()) != null)
				{
					string textToAdd = line;
					if (line.StartsWith("."))
					{
						textToAdd = line.Insert(0, ".");
					}
					sb.AppendLine(textToAdd);
				}
			}
		}
                // write modified stream to disk
		using (FileStream fileStream = new FileStream(path, FileMode.CreateNew))
		{
			StreamWriter streamWriter = new StreamWriter(fileStream);
			streamWriter.Write(sb.ToString());
			streamWriter.Close();
			fileStream.Close();
		}
		return;
	}
	catch (IOException)
	{
		Thread.Sleep(100);
		tryCount += 1;
		// The file may have been created between our File.Exists() check and
		// our attempt to create the stream.
	}
} while (tryCount < 100);

// Couldn't write to disk for some reason. (probably drive full. probably not filename collision)
// Throw away the message and reset the waiting email record so we pick it up sometime later
if(tryCount >= 100)
{
	  //cleanup/retry later code.
}
}
```

**Comment by jstedfast** [comment]
Thanks for the info, I did not realize that messages in IIS's pickup folder needed to be SMTP dot-stuffed.

FWIW, you can feel free to grab MailKit's [SmtpDataFilter.cs](https://github.com/jstedfast/MailKit/blob/master/MailKit/Net/Smtp/SmtpDataFilter.cs) code which will do the dot-stuffing for you.

You can use it like this:

```csharp
using (var file = File.Create ("fileName.eml")) {
    using (var filtered = new FilteredStream (file)) {
        filtered.Add (new SmtpDataFilter ());

        // make sure MimeKit canonicalizes in DOS format
        var options = FormatOptions.Default.Clone ();
        options.NewLineFormat = NewLineFormat.dos;

        message.WriteTo (options, filtered);
        filtered.Flush ();
    }
}
```

I'll clean this up for the FAQ when I get time (in the middle of work atm).

**Comment by jstedfast** [comment]
Here's the code snippet I came up with. I'll probably need to make SmtpDataFilter public (it is currently internal) before I publish this update to the docs:

```csharp
void SendToPickupDirectory (MimeMessage message, string pickupDirectory)
{
    do {
        var path = Path.Combine (pickupDirectory, Guid.NewGuid ().ToString () + ".eml");

        // if the file already exists, then try again with a new Guid.
        if (File.Exists (path))
            continue;

        bool created = false;

        try {
            using (var stream = new File.Create (path)) {
                // keep track f whether or not we've created the .eml file
                created = true;

                // IIS pickup directories expect the message to be "byte-stuffed"
                // which means that lines beginning with "." need to be escaped
                // by adding an extra "." to the beginning of the line.
                //
                // Let MimeKit.IO.FilteredStream with MailKit.Net.Smtp.SmtpDataFilter
                // do this "byte-stuffing" for us the same way that SmtpClient.Send()
                // would do it.
                using (var filtered = new FilteredStream (stream)) {
                    filtered.Add (new SmtpDataFilter ());

                    // make sure to write the message in DOS (<CR><LF>) format.
                    var options = FormatOptions.Default.Clone ();
		    options.NewLineFormat = NewLineFormat.Dos;

                    message.WriteTo (options, stream);
                    return;
                }
            }
        } catch (IOException ex) {
            if (created) {
                // clean up the file we've created
                File.Delete (path);

                // re-throw the exception - this may mean that the disk is full
		throw;
            }

            // if we didn't succeed in creating the file, then this IOException
	    // probably means that the file was created by another process between
	    // File.Exists() and File.Freate().
        }
    } while (true);
}
```

**Comment by jstedfast** [comment]
Hmmm, maybe this is a bit cleaner...

```csharp
void SaveToPickupDirectory (MimeMessage message, string pickupDirectory)
{
    do {
        // Generate a random file name to save the message to.
        var path = Path.Combine (pickupDirectory, Guid.NewGuid ().ToString () + ".eml");
        Stream stream;

        try {
            // Attempt to create the new file.
            stream = File.Open (path, FileMode.CreateNew);
        } catch (IOException) {
	    // If the file already exists, try again with a new Guid.
            if (File.Exists (path))
                continue;

            // Otherwise, fail immediately since it probably means that there is
            // no graceful way to recover from this error.
            throw;
        }

        try {
            using (stream) {
                // IIS pickup directories expect the message to be "byte-stuffed"
                // which means that lines beginning with "." need to be escaped
                // by adding an extra "." to the beginning of the line.
                //
                // Use an SmtpDataFilter "byte-stuff" the message as it is written
                // to the file stream. This is the same process that an SmtpClient
                // would use when sending the message in a `DATA` command.
                using (var filtered = new FilteredStream (stream)) {
                    filtered.Add (new SmtpDataFilter ());

                    // Make sure to write the message in DOS (<CR><LF>) format.
                    var options = FormatOptions.Default.Clone ();
                    options.NewLineFormat = NewLineFormat.Dos;

                    message.WriteTo (options, filtered);
                    filtered.Flush ();
                    return;
                }
            }
        } catch {
            // An exception here probably means that the disk is full.
            //
            // Delete the file that was created above so that incomplete files are not
            // left behind for IIS to send accidentally.
            File.Delete (path);
            throw;
        }
    } while (true);
}
```

**Comment by jstedfast** [comment]
Fixed in commit aec28ffb99b9696e1b8ad81df93155b542b143ef

**Comment by jstedfast** [comment]
SmtpDataFilter is now available in 2.4.1
